Added ImGui panel to mqsettings, changes, and fixes applied.

hytiek opened the pull request:
Added
- MQ Settings panel (/mqsettings -> plugin -> AutoSize)
- AutoSize TLO (check wiki for details)
- Introduced Synchronization ability - only available in the UI
- Enhanced options to accept instruction such as on and off, not just be a toggle. This was done to enable keeping toons in sync by setting values instead of randomly toggling.

Updated / Fixed
- Zonewide now uses 1000 range
- /autosize range # now works
- Reduced organic growth of code and made all toggle options use the ToggleOption function
- Fixed and altered how "Everything" works. Now it will enable the options, which work as intended.

Deprecated
- OptZone since it never worked as expected, replaced with Range of 1000 which is about the same as the EQ visible max clipping plane
- ResizeAll configuration, while Everything is still able to be enabled
- DefaultSize configuration and command (/autosize size), while Everything is still able to be enabled
- SizeDefault since having a default setting to resize something to, which wasn't opted in for resizing, doesn't actually make any sense. Outputs have persisted to ensure no script that scrapes the line breaks.
- SizeTarget (and /autosize target) since it would only resize for one frame if the type was managed by an Option for resizing
- OptByRange and SizeByRange, since the plugin now only uses range to resize


Under the hood
- Changed the majority of floats to integer as there was no reason for using all the floats
- Updated pCharSpawn to pLocalPlayer references

Commit:
Added ImGui panel to mqsettings and bunch of changes and fixes

Added MQ Settings panel (/mqsettings -> plugin -> AutoSize)
Updated Zonewide to work (temporarily changes range size to 1000)
Fixed /autosize range # to work
Replaced "AllZone" with "Zonewide"
Reduced organic growth of code and made all toggle options use the ToggleOption function
Changed the majority of floats to integer as there was no reason for using all the floats.
Changed how Everything works to rather enable all the options as prior implementation didn't work as it should have
Deprecated the use of pChSpawn, updated to use pLocalPlayer

Commit:
adding TLO data

Commit:
simplified the panel UI

Commit:
prior to deprecating OptRange and SizeByRange

Commit:
fixed default setting for optZonewide, deprecated SizeTarget and cmd

Commit:
deprecated things which related to "Everything"

Commit:
deprecated SizeDefault

Deprecated SizeDefault since having a default setting to resize something to, which wasn't opted in for resizing, doesn't actually make any sense. Outputs have persisted to ensure no script that scrapes the line breaks.

Commit:
enhanced each toggle to allow for on | off setting

This will be used with the synchronization buttons

Commit:
fixed zone vs range and added some group commands

Commit:
another fix for zonewide vs range

also removed references of 1000 and replaced with use of FAR_CLIP_PLANE

Commit:
disabled None from sync if either comms are loaded

Commit:
new SaveINI, added round10 func, updated sliders for input

new SaveINI which allows for squelch and param (key) to be passed as wanted.

Commit:
wrapped the SaveINI per key with a check if autosave is enabled

Commit:
fixed typo for NPC, added wrapper for saving on unload

It makes no sense to have temporary settings if while temporarily in play until you camp out and exit the plugin. At that point the original would save those "temp" values. That defeats the whole purpose for temporary settings.

Commit:
refactored how instructions were generated

Commit:
updated comments

Commit:
added isInGroup, isInRaid for button control

Commit:
removed debug message

Commit:
fixed clip plane instruction when autosave is enabled

Commit:
removed lingering TODO: delete after testing

Commit:
removed double slash from existing comment

Commit:
cleaned up comments, ordered definitions, updated 2 functions

updated emulate and SendGroupCommand

Commit:
Updated description of plugin

Commit:
comment and reference clean up

Commit:
changed how to disable the range slider

Commit:
removed empty lines

Commit:
initial changelog entry

hytiek committed:
Merge pull request #1 from hytiek/imgui_panel_and_fixes

Modernizing MQ2AutoSize and fixes

grimmier378 commented:
I've been testing this as Hytiek has been modifying it and, helping him debug any issues with the plugin.  Since the last iteration I haven't found anything not working, and more things that do work as they were intended or at least how one would expect them to function. Great work Hytiek.

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -1,627 +1,1387 @@
-// v1.0 - Eqmule 07-22-2016 - Added string safety.
-//
-//////////////////////////////////////////////////////////////////////////////
-// Usage:
-//	This plugin will automatically resize configured spawns to the specified
-//	size. You can configure it to only resize within a specific range and then
-//	resize back to normal when your distance moves out of that range.
-//	Current default range is set to 50 and may be changed via INI or cmd line
-//	NOTE:  These effects are CLIENT SIDE ONLY!
-//
-// Commands:
-//  /autosize				  - Toggles zone-wide  AutoSize on/off
-//  /autosize dist			- Toggles distance-based AutoSize on/off
-//  /autosize pc			  - Toggles AutoSize PC spawn types
-//  /autosize npc			 - Toggles AutoSize NPC spawn types
-//  /autosize pets			- Toggles AutoSize pet spawn types
-//  /autosize mercs		  - Toggles AutoSize mercenary spawn types
-//  /autosize mounts		 - Toggles AutoSize mounted player spawn types
-//  /autosize corpse		 - Toggles AutoSize corpse spawn types
-//  /autosize target		 - Resizes your target to sizetarget size
-//  /autosize everything	- Toggles AutoSize all spawn types
-//  /autosize self			- Toggles AutoSize for your character
-//  /autosize range #		- Sets range for distance-based AutoSize
-//
-//  (Valid sizes 1 to 250)
-//  /autosize size #		 - Sets default size for "everything"
-//  /autosize sizepc #	  - Sets size for PC spawn types
-//  /autosize sizenpc #	 - Sets size for NPC spawn types
-//  /autosize sizepets #	- Sets size for pet spawn types
-//  /autosize sizetarget # - Sets size for target parameter
-//  /autosize sizemercs #  - Sets size for mercenary spawn types
-//  /autosize sizemounts # - Sets size for mounted player spawn types
-//  /autosize sizecorpse # - Sets size for corpse spawn types
-//  /autosize sizeself #	- Sets size for your character
-//
-//  /autosize status		 - Display current plugin settings to chatwnd
-//  /autosize help			- Display command syntax to chatwnd
-//  /autosize save			- Save settings to INI file (auto on plugin unload)
-//  /autosize load			- Load settings from INI file (auto on plugin load)
-//  /autosize autosave	  - Automatically save settings to INI file when an option is toggled or size is set
```
This seems like good material for the readme, which is currently empty (I don't know how accurate it is - didn't read it, but just deleting it seemed like we could do better...)

hytiek replied:
This information was moved to the imgui panel, I have added it to the README.md

hytiek replied:
![image](https://github.com/user-attachments/assets/a56cf82d-e990-41a6-a964-b82bb96fc0d2)

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -1,627 +1,1387 @@
 #include <mq/Plugin.h>
 
+// Constants
 const char* MODULE_NAME = "MQ2AutoSize";
```
`mqplugin::PluginName` exists

hytiek replied:
I've made the change as suggested.

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -1,627 +1,1387 @@
 
+// Constants
 const char* MODULE_NAME = "MQ2AutoSize";
+const int SKIP_PULSES = 5;               // Controls the number of pulses to perform a radius-based resize
+const int MIN_SIZE = 1;                  // Minimum size value
+const int MAX_SIZE = 250;                // Maximum size value
+int FAR_CLIP_PLANE = 1000;               // Placeholder for EQ Far Clip Plane value
+const float OTHER_SIZE = 1.0f;           // Default size for other entities
+const float ZERO_SIZE = 0.0f;            // Size representing zero
+const int DEFAULT_COMMS = 0;             // Default comms value
+const int DEFAULT_OPT_ZONEWIDE = 2;      // Default option for zonewide
```
lets make these constexpr, and should FAR_CLIP_PLANE be a constexpr as well?

hytiek replied:
I've applied this and included FAR_CLIP_PLANE.

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -1,627 +1,1387 @@
+int optZonewide = DEFAULT_OPT_ZONEWIDE;  // Option for zonewide, defaults to selecting Range
+bool loaded_dannet = false;              // State of DanNet plugin
+bool loaded_eqbc = false;                // State of EQBC plugin
+unsigned long long commsCheck;           // Comms check timestamp
+unsigned long long commsRefreshRateSeconds = 1;	// Refresh rate (seconds) used in OnPulse to check if comms plugins are available
```
in general: 
unsigned long long -> uint64_t
long long -> int64_t

hytiek replied:
updated, will be in the next push

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -1,627 +1,1387 @@
+	bool OptNPC;
+	bool OptPet;
+	bool OptMerc;
+	bool OptMount;
+	bool OptCorpse;
+	bool OptSelf;
 
 	int ResizeRange;
-
-	float SizeDefault;
-	float SizePC;
-	float SizeNPC;
-	float SizePet;
-	float SizeMerc;
-	float SizeTarget;
-	float SizeMount;
-	float SizeCorpse;
-	float SizeSelf;
+	int SizePC;
+	int SizeNPC;
+	int SizePet;
+	int SizeMerc;
+	int SizeMount;
+	int SizeCorpse;
+	int SizeSelf;
```
since you're changing this anyways, i would suggest moving these all to member initializers

```
bool OptAutoSave = false;
bool OpcPC = true
// ... etc
int SizePC = 1;
```
etc

hytiek replied:
updated, will be in the next push

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -1,627 +1,1387 @@
 COurSizes AS_Config;
 
+// exposed TLO variables
+class MQ2AutoSizeType* pAutoSizeType = 0;
```
put this after the class so you don't need to forward declare MQ2AutoSize (can remove "class" from this declaration)

hytiek replied:
updated, will be in the next push

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -1,627 +1,1387 @@
+class MQ2AutoSizeType* pAutoSizeType = 0;
+class MQ2AutoSizeType : public MQ2Type {
+public:
+	enum AutoSizeMembers {
```
enum class

hytiek replied:
updated

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -1,627 +1,1387 @@
+	};
+
+	MQ2AutoSizeType() :MQ2Type("AutoSize") {
+		TypeMember(Active);
```
once you make those enum class, these can use ScopedTypeMember

hytiek replied:
updated

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -1,627 +1,1387 @@
+		SizeSelf
+	};
+
+	MQ2AutoSizeType() :MQ2Type("AutoSize") {
```
```suggestion
	MQ2AutoSizeType() : MQ2Type("AutoSize")
	{
```

hytiek replied:
updated

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -1,627 +1,1387 @@
+	}
+
+	bool ToString(MQVarPtr VarPtr, char* Destination) {
+		return true;
```
this shouldn't return true without assigning a value to Destination

hytiek replied:
updated

brainiac replied:
since it doesn't really do anything and behaves in a well defined manner. However we could also just not have this function at all. But this is fine, you don't need to change it, just want to point out the option.

brainiac commented:
I will finish my review later, but for now, for the love of all things nice and beautiful, put the braces back, lol.

hytiek commented:
Made things beautiful again.

Commit:
Provided { with VIP status, constexpr added, updated readme

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -1,111 +1,231 @@
+
+		switch ((AutoSizeMembers)pMember->ID)
+		{
+			case Active:
```
... and if you use scoped enum, then these would be AutoSizeMembers::Active

hytiek replied:
updated

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -131,88 +251,125 @@ class PlayerZoneClient_Hook
-	if (fValidate > MAX_SIZE)
-	{
-		return MAX_SIZE;
-	}
-	return fValidate;
+	char szTemp[MAX_STRING] = { 0 };
+	GetPrivateProfileString(section, key, defaultValue, szTemp, MAX_STRING, INIFileName);
+	return (ci_equals(szTemp, "on"));
```
This function can be replaced with GetPrivateProfileBool

hytiek replied:
updated

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -131,88 +251,125 @@ class PlayerZoneClient_Hook
-	AS_Config.SizeCorpse = SaneSize((float)GetPrivateProfileInt("Config", "SizeCorpse", (int)MIN_SIZE, INIFileName));
-	AS_Config.SizeSelf = SaneSize((float)GetPrivateProfileInt("Config", "SizeSelf", (int)MIN_SIZE, INIFileName));
-
-	// apply new INI read
-	if (GetGameState() == GAMESTATE_INGAME && pLocalPlayer && AS_Config.OptByZone) SpawnListResize(false);
+	int size = GetPrivateProfileInt(section.c_str(), key.c_str(), defaultValue, INIFileName);
+	return std::clamp(size, MIN_SIZE, MAX_SIZE);
```
This function accepts a std::string, but then only uses c_str() on it. That means if you have a const char* to begin with, it will convert to std::string and then convert back to const char* through c_str().

It would be better to just accept a const char* in this function, because there is no need to utilize a std::string here.

hytiek replied:
updated

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -131,88 +251,125 @@ class PlayerZoneClient_Hook
-	WritePrivateProfileString("Config", "SizeCorpse", SafeItoa((int)AS_Config.SizeCorpse, szTemp, 10), INIFileName);
-	WritePrivateProfileString("Config", "SizeSelf", SafeItoa((int)AS_Config.SizeSelf, szTemp, 10), INIFileName);
+	// Map to store configuration key-value pairs
+	std::map<std::string, std::string> configMap =
```
you're not really leveraging any of the strengths of a std::map here, so I would recommend instead using `std::vector<std::pair<std::string, std::string>>` - a vector of pairs.

The only thing you'd need to change is the configMap.find to
```c++
auto it = std::find_if(configMap.begin(), configMap.end(), [&](const auto& p) { return p.first == param; });
```

hytiek replied:
updated

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -225,10 +382,13 @@ void ChangeSize(PlayerClient* pChangeSpawn, float fNewSize)
+
 	PSPAWNINFO pLPlayer = (PSPAWNINFO)pLocalPlayer;
-	if (!pLPlayer || !pChSpawn->SpawnID || !pSpawn || !pSpawn->SpawnID) return;
+	if (!pLPlayer || !pLPlayer->SpawnID || !pSpawn || !pSpawn->SpawnID) return;
```
this can use pLocalPlayer directly instead of pLPlayer. Also don't need to check SpawnID. Thats pretty overly defensive.

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -225,10 +382,13 @@ void ChangeSize(PlayerClient* pChangeSpawn, float fNewSize)
 void SizePasser(PSPAWNINFO pSpawn, bool bReset)
 {
-	if ((!bReset && !AS_Config.OptByZone && !AS_Config.OptByRange) || GetGameState() != GAMESTATE_INGAME) return;
-	PSPAWNINFO pChSpawn = (PSPAWNINFO)pCharSpawn;
```
Unsure if this is intentional, but pCharSpawn is pControlledPlayer, not pLocalPlayer.

pControlledPlayer is the player that you're currently in control of. So really this will be different for an eye of zomm, or maybe a mount.

hytiek replied:
The original plugin did not resize a controlled spawn like Eye of Zomm. Mounts were handled successfully in the original, and continue to resize appropriately.

I've been unsuccessful in trying to get Eye of Zomm to resize using pControlledPlayer.

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -238,61 +398,61 @@ void SizePasser(PSPAWNINFO pSpawn, bool bReset)
-	PSPAWNINFO pChSpawn = (PSPAWNINFO)pCharSpawn;
 	PSPAWNINFO pLPlayer = (PSPAWNINFO)pLocalPlayer;
-	if (GetGameState() != GAMESTATE_INGAME || !pLPlayer || !pChSpawn->SpawnID || !pSpawn || !pSpawn->SpawnID) return;
+	if (GetGameState() != GAMESTATE_INGAME || !pLPlayer || !pLPlayer->SpawnID || !pSpawn || !pSpawn->SpawnID)
```
don't need pLPlayer. Also don't really need the SpawnID checks.

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -331,13 +497,24 @@ PLUGIN_API void OnEndZone()
+	// check if communication plugins are still running and adjust UI as needed
+	if (GetTickCount64() > commsCheck)
+	{
+		commsCheck = std::int64_t(commsCheck) + (static_cast<long long>(commsRefreshRateSeconds) * 1000);
```
not really sure what is going on here with the casts, but if you fix the types as i mentioned at the top of the file you shouldn't need these.

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -359,269 +536,1087 @@ PLUGIN_API void OnPulse()
+	if (!AS_Config.OptPC && !AS_Config.OptNPC && !AS_Config.OptPet && !AS_Config.OptMerc && !AS_Config.OptCorpse && !AS_Config.OptSelf)
 	{
-		sprintf_s(szMethod, "\ayZonewide\ax");
+		sprintf_s(szMethod, "\arInactive\ax");
```
strcpy_s when there is no format string.

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -359,269 +536,1087 @@ PLUGIN_API void OnPulse()
 	{
-		sprintf_s(szMethod, "\arInactive\ax");
+		// covers the idea of Zonewide by using FAR_CLIP_PLANE value (which is about 100% Far Cliping Plane)
+		sprintf_s(szMethod, "\ayZonewide\ax");
```
strcpy_s when there is no format string.

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -359,269 +536,1087 @@ PLUGIN_API void OnPulse()
 	GetArg(szCurArg, szLine, 1);
 	GetArg(szNumber, szLine, 2);
-	float fNewSize = (float)atof(szNumber);
+	int iNewSize = std::atoi(szNumber);
```
use GetIntFromString instead.

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -359,269 +536,1087 @@ PLUGIN_API void OnPulse()
-		SaveINI();
-}
-
 void AutoSizeCmd(PSPAWNINFO pLPlayer, char* szLine)
```
This should be `void AutoSizeCmd(PlayerClient* pLPlayer, const char* szLine)`

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -359,269 +536,1087 @@ PLUGIN_API void OnPulse()
-			return;
 		}
-		else if (!_strnicmp(szCurArg, "save", 5))
+		else if (!ci_equals(szNumber, "on") && !ci_equals(szNumber, "off"))
```
if condition not really needed here given the previous conditions.

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -359,269 +536,1087 @@ PLUGIN_API void OnPulse()
+			}
 		}
-		else if (!_strnicmp(szCurArg, "autosave", 9))
+		else if (!ci_equals(szNumber, "on") && !ci_equals(szNumber, "off"))
```
same as above. You can't get here if szNumber was on or off, so condition isn't necessary.

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -359,269 +536,1087 @@ PLUGIN_API void OnPulse()
+	{
+		if (!AS_Config.OptSelf)
+		{
+			DoCommandf("/squelch /autosize self");
```
You're not using a format string. use DoCommand instead for all of these.

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -359,269 +536,1087 @@ PLUGIN_API void OnPulse()
+			ResetAllByType(PET);
 		}
-		else if (!_strnicmp(szCurArg, "sizecorpse", 11))
+		else if (!ci_equals(szNumber, "on") && !ci_equals(szNumber, "off"))
```
unnecessary condition. this keeps repeating, i'm not going to comment on any further ones.

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -359,269 +536,1087 @@ PLUGIN_API void OnPulse()
+				AS_Config.OptAutoSave = !AS_Config.OptAutoSave;
+				DoCommandf("/autosize autosave");
+			}
+			// General: Zodewide
```
typo in comment

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -359,269 +536,1087 @@ PLUGIN_API void OnPulse()
+				ImGui::TableSetupColumn("", ImGuiTableColumnFlags_WidthFixed, 20.0f);
+				ImGui::TableSetupColumn("", ImGuiTableColumnFlags_WidthFixed);
+				ImGui::TableNextColumn();
+				if (ImGui::RadioButton("", &optZonewide, static_cast<int>(ResizeMode::Range)))
```
This could use an id

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -359,269 +536,1087 @@ PLUGIN_API void OnPulse()
+				ImGui::TableNextColumn();
+				ImGui::SetNextItemWidth(50.0f);
+				ImGui::BeginDisabled(!AS_Config.OptPC);
+				if (ImGui::SliderInt("Resize: Other player(s) (incluldes those mounted)##inputOP", &AS_Config.SizePC, 1, 30, "%d", ImGuiSliderFlags_NoInput | ImGuiSliderFlags_AlwaysClamp))
```
The label is used as an Id, if the label is unique you don't need the ##stuff after it.

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -359,269 +536,1087 @@ PLUGIN_API void OnPulse()
+			ImGui::BeginDisabled(loaded_dannet || loaded_eqbc);
+			if (ImGui::RadioButton("None", &selectedComms, static_cast<int>(CommunicationMode::None)))
+			{
+				selectedComms = static_cast<int>(CommunicationMode::None);
```
Assigning the value is unnecessary. the RadioButton will do it for you.

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -359,269 +536,1087 @@ PLUGIN_API void OnPulse()
+				ImGui::TableNextColumn();
+				if (ImGui::Checkbox("##OptPet", &AS_Config.OptPet))
+				{
+					AS_Config.OptPet = !AS_Config.OptPet;
```
Assigning the value is unnecessary. The Checkbox will do it for you (thats why you pass its address)

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -359,269 +536,1087 @@ PLUGIN_API void OnPulse()
+					DoCommandf("/autosize corpse");
+				}
+				ImGui::TableNextColumn();
+				ImGui::PushItemWidth(50.0f);
```
Seeing a ton of PushItemWidth and no PopItemWidth.

Probably better off using SetNextItemWidth instead.

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -359,269 +536,1087 @@ PLUGIN_API void OnPulse()
 
-	AddCommand("/autosize", AutoSizeCmd);
-	LoadINI();
+	loaded_eqbc = GetPlugin("MQ2EQBC");
```
IsPluginLoaded

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -331,13 +497,24 @@ PLUGIN_API void OnEndZone()
+	if (GetTickCount64() > commsCheck)
+	{
+		commsCheck = std::int64_t(commsCheck) + (static_cast<long long>(commsRefreshRateSeconds) * 1000);
+		ChooseInstructionPlugin();
```
Oh geeze, i see what you're doing.

Don't do this. Just Implement OnLoadPlugin/OnUnloadPlugin plugin callbacks

```
PLUGIN_API void OnLoadPlugin(const char* pluginName)
{
    if (ci_equals(pluginName, "MQ2EQBC"))
    {
        // ... we just loaded MQ2EQBC
    }
}
```

etc

hytiek replied:
To work correctly I need the change I shared with you on Discord to be implemented in MQPluginHandler.cpp. I'll resolve this once that change is available.

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -359,269 +536,1087 @@ PLUGIN_API void OnPulse()
+ */
+// this function is used as a toggle between Zone and Range
+// params: zonewide or range
+void emulate(const std::string& type)
```
you only ever pass string literals to this function consider using std::string_view instead.

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -1,111 +1,231 @@
+char szTemp[MAX_STRING] = { 0 };         // Temporary buffer for strings
+int previousRangeDistance = 0;           // Previous range distance
+int selectedComms = DEFAULT_COMMS;       // Selected comms, updated in OnPulse
+int optZonewide = DEFAULT_OPT_ZONEWIDE;  // Option for zonewide, defaults to selecting Range
```
Maybe this should be type ResizeMode?

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -1,111 +1,231 @@
+unsigned int uiSkipPulse = 0;            // Skip pulse counter
+char szTemp[MAX_STRING] = { 0 };         // Temporary buffer for strings
+int previousRangeDistance = 0;           // Previous range distance
+int selectedComms = DEFAULT_COMMS;       // Selected comms, updated in OnPulse
```
Maybe this should be type CommunicationMode?

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -1,111 +1,231 @@
+constexpr const float OTHER_SIZE = 1.0f;           // Default size for other entities
+constexpr const float ZERO_SIZE = 0.0f;            // Size representing zero
+constexpr const int DEFAULT_COMMS = 0;             // Default comms value
+constexpr const int DEFAULT_OPT_ZONEWIDE = 2;      // Default option for zonewide
```
These should be values on the enum.

```c++
enum Foo
{
    a = 1,
    b = 2,
    c = 3,

    Default = b
}
```

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -359,269 +536,1087 @@ PLUGIN_API void OnPulse()
+}
 
+// check if we are in a group
+static bool isInGroup()
+{
+	if (pLocalPC)
+	{
+		if (pLocalPC->Group)
+		{
+			return true;
+		}
+	}
+	return false;
```
`return pLocalPC && pLocalPC->Group`

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -359,269 +536,1087 @@ PLUGIN_API void OnPulse()
 }
+
+// check if we are in a raid
+static bool isInRaid()
+{
+	if (pRaid)
+	{
+		if (pRaid->RaidMemberCount)
+		{
+			return true;
+		}
+	}
+	return false;
```
`return pRaid && pRaid->RaidMemberCount`

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -359,269 +536,1087 @@ PLUGIN_API void OnPulse()
+	}
+	else if (value > MAX_SIZE)
+	{
+		return 250;
```
using 250 here instead of MAX_SIZE constant.

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -359,269 +536,1087 @@ PLUGIN_API void OnPulse()
+// send instruction to selected groups:
+// -> DanNet: all, zone, raid, group
+// -> EQBC: all, group
+void SendGroupCommand(const std::string& who)
```
consider changing this to a std::string_view

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -359,269 +536,1087 @@ PLUGIN_API void OnPulse()
+{
+	ImGui::TextColored(ImVec4(1.0f, 1.0f, 0.0f, 1.0f), "MQ2AutoSize");
+	ImGuiTabBarFlags tab_bar_flags = ImGuiTabBarFlags_None;
+	if (ImGui::BeginTabBar("AutoSizeTabBar", tab_bar_flags))
```
flags is optional and you're not providing any, so feel free to omit.

brainiac reviewed (commented):
Left a bunch of comments, hope this helps.

Also consider searching for every reference to PSPAWNINFO and doing something with it. PSPAWNINFO casts are unnecessary, and any variables of the type should be PlayerClient* instead.

Commit:
addressed all points from pull/8

this update includes all things related to https://github.com/RedGuides/MQ2AutoSize/pull/8

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -1,110 +1,224 @@
+	DanNet = 1,
+	EQBC = 2,
+
+	Default = 0
```
You can write `Default = None` here for clarity. Same idea below with eResizeMode: `Default = Range`

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -1,110 +1,224 @@
+
+	Default = 0
+};
+int selectedComms = static_cast<int>(eCommunicationMode::Default);
```
So here is a tip. If you change this from `int` to `eCommunicationMode selectedComms` then you can remove all your casts.

Then write this helper function and use it for your radio buttons involving the enums. It'll keep everything typesafe and then you aren't making friction against your enums with all the casting to int.

```c++
template <typename T>
bool RadioButton(const char* label, T* value, T defaultValue)
{
	return ImGui::RadioButton(label, static_cast<std::underlying_type<T>*>(value),
		static_cast<std::underlying_type<T>>(defaultValue));
}
```

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -1,110 +1,224 @@
 public:
-	COurSizes()
+	COurSizes() = default;
+	bool OptAutoSave = 0;
+	bool OptPC = 1;
+	bool OptNPC = 0;
+	bool OptPet = 0;
+	bool OptMerc = 0;
+	bool OptMount = 0;
+	bool OptCorpse = 0;
+	bool OptSelf = 0;
```
bools should be initialized with true or false.

brainiac commented on MQ2AutoSize.cpp:
```diff
@@ -359,269 +509,1110 @@ PLUGIN_API void OnPulse()
 
+PLUGIN_API void OnLoadPlugin(const char* pluginName)
+{
+	if (ci_equals(pluginName, "MQ2DanNet"))
+	{
+		
+	}
+
+	if (ci_equals(pluginName, "MQ2EQBC"))
+	{
+
+	}
```
it doesn't look like you finished this

Commit:
additional updated from pull/8

Commit:
updated how selectedComms is defined

Commit:
updated ResizeMode to better use enum

hytiek commented:
All comments have been resolved
